### PR TITLE
feat(apps): add apps release workflow

### LINKED ISSUE #943: `/apps/ocean-api`: CD Publish Docker Image
> As part of #580 consolidation efforts. 
> https://github.com/DeFiCh/whale/tree/main/.github/workflows

#### What would you like to be added:

On "Release", "PR into main" and "merge into main"; build and publish a docker image of `apps/ocean-api`.

- Release: Tagged with `release version` and `latest`
- PR into main: Tagged with `commit hash` and `main`
- PR for merge into main: Tagged with `commit hash` and `pr-{number}`

#### Why is this needed:

Shift left.

/sig ecosystem
/area apps
/triage accepted

### PR DESCRIPTION
**Pull request by kodemon**
#### What this PR does / why we need it:

Add shift left automation for application containers.

#### Which issue(s) does this PR fixes?:

Fixes #943

#### Additional comments?:

This is part 2 following up on

 - https://github.com/DeFiCh/jellyfish/pull/1085

This consolidates all application release logic into a single yml workflow file, and adds staging builds when merging to the main branch and dev builds when creating PRs against main.

### CODE REVIEW

**Review by canonbrother — DISMISSED** [review]

**Comment by kodemon** [comment]
I want to emphasize that the build times for our docker containers are ~1 hour long because of the `arm64` requirement. This workflow triggers on release, main and PR to main. I think for release and main the build time should be fine. But for PRs they may be a little intrusive, so depending on how things goes we may want to be more selective with the PR triggers.

We can be more targeted I think, maybe do manual build triggers for PRs so that we can opt in on container releases for branches that see a lot of activity?

@fuxingloh thoughts?

**Comment by fuxingloh** [comment]
> @fuxingloh thoughts?

I think it's fine because we don't need to wait for them to pass before merging - furthermore our PR takes way longer than 1 hour to review.

By taking on the shift left mentality, we can have visibility of failures earlier which is good!

File: .github/scripts/release-tags.js
Comment (fuxingloh) [inline .github/scripts/release-tags.js:60]:
```suggestion
  const sha = context.sha.substr(0, 12)
  return `ghcr.io/defich/${app}:main,ghcr.io/defich/${app}:main-${sha}`
```

File: .github/workflows/release-apps.yml
Comment (fuxingloh) [inline .github/workflows/release-apps.yml:52]: For PR we should add something like this, easily allow the reader to grab the hash for testing.

via: `if: github.event_name == 'pull_request'`

https://github.com/DeFiCh/wallet/blob/df1696ccb922541c9b279944abb39d3bf4ee8b10/.github/workflows/expo-preview.yml#L35-L44

**Comment by fuxingloh** [comment]
A few minor QOL suggestions.

File: .github/workflows/release-apps.yml
Comment (kodemon) [reply .github/workflows/release-apps.yml:52]: yeah, that makes sense, will update.

File: .github/scripts/release-tags.js
Comment (kodemon) [reply .github/scripts/release-tags.js]: updated, since we are using it in two places I made a `getShortSha` method

File: .github/workflows/release-apps.yml
Comment (canonbrother) [inline .github/workflows/release-apps.yml:20]:
```suggestion
      - uses: actions/checkout@230611dbd0eb52da1e1f4f7bc8bb0c3a339fc8b7
```
better
https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions

File: .github/scripts/release-tags.js
Comment (fuxingloh) [reply .github/scripts/release-tags.js]: To add another `ghcr.io/defich/${app}:main` too

Comment (kodemon) [reply .github/scripts/release-tags.js]: done, I actually had the tagging a little wrong so I have updated to adhere to the request in the ticket

File: .github/workflows/release-apps.yml
Comment (fuxingloh) [inline .github/workflows/release-apps.yml:8]:
```suggestion
  pull_request:
```

I think it should be `pull_request`, since `_target` is a privileged environment.

Comment (kodemon) [reply .github/workflows/release-apps.yml]: changed to use `pull_request`

File: .github/workflows/release-apps.yml
Comment (kodemon) [inline .github/workflows/release-apps.yml:70]: This is not yet outputting the way we want, still working on this.

File: .github/workflows/release-apps.yml
Comment (fuxingloh) [inline .github/workflows/release-apps.yml:59]: Minor suggestion for optimization, since you are checking out the entire workflow with this step. Would it make sense to just have GitHub scripts with the embedded script since it's pretty short - to also keep it close and contextual together?

Just a suggestion, I am on the fence here. Either way works imo.

File: .github/workflows/release-apps.yml
Comment (fuxingloh) [inline .github/workflows/release-apps.yml:10]: For security hardening should add permission.

https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs
https://docs.github.com/en/rest/reference/permissions-required-for-github-apps

File: .github/workflows/release-apps.yml
Comment (fuxingloh) [inline .github/workflows/release-apps.yml:12]: `if: github.event_name != 'pull_request' || github.event.pull_request.head.repo.full_name == github.repository`

To prevent all workflow from running if it's coming from a forked pull request. Since it can't run without failing as they can't get the access to write into ghcr.io

**Review by fuxingloh — CHANGES_REQUESTED** [review]
A few more minor suggestions and changes

File: .github/workflows/release-apps.yml
Comment (kodemon) [reply .github/workflows/release-apps.yml:59]: While I would definitely advocate keeping things as close as possible, I think it this case it bloats the file too much. If the scripts themselves were smaller i'd probably leave them in though. In this case though I have emphasized writing it up so that any future additions to the yml stays isolated to the yml while the scripts are more abstract.

I haven't seen any performance concerns with the checkout yet, but if we see it as an issue we could move the scripts inline.

I will defer to your judgment here, but personally I think its at least more readable this way.

File: .github/workflows/release-apps.yml
Comment (kodemon) [reply .github/workflows/release-apps.yml:12]: After a little thought I updated the script to use the github org/username as deploy target so that we don't have to worry about hard coded endpoints when people fork the repository.

File: .github/workflows/release-apps.yml
Comment (kodemon) [reply .github/workflows/release-apps.yml:10]: I have been wondering what sort of permissions we may want to add here.

I assume we add permissions so that any third party package we are using is not able to perform operations on our repository they are not permitted to?

If my assumption is correct I will work to figure out what the minimum permissions are to fulfill our workflow.

File: .github/workflows/release-apps.yml
Comment (fuxingloh) [inline .github/workflows/release-apps.yml:50]: This should be separated into a matrix because we want to run them concurrently so that it's faster in parallel.

For speed to deployment - we should not need to wait for both platforms to be built and then pushed.

File: .github/workflows/release-apps.yml
Comment (fuxingloh) [inline .github/workflows/release-apps.yml:29]:
```suggestion
        uses: docker/login-action@bb984efc561711aaa26e433c32c3521176eae55b
```

A recent update in https://github.com/DeFiCh/jellyfish/pull/1123

File: .github/workflows/release-apps.yml
Comment (kodemon) [reply .github/workflows/release-apps.yml:50]: Makes sense, will update.

File: .github/workflows/release-apps.yml
Comment (fuxingloh) [reply .github/workflows/release-apps.yml:10]: To create another PR for this, since it's low priority.

Comment (fuxingloh) [reply .github/workflows/release-apps.yml:10]: Created a new ticket to track it: https://github.com/DeFiCh/jellyfish/issues/1151

**Comment by fuxingloh** [comment]
@kodemon can you create another PR to add `playground-api` in the matrix?

**Pull request merged**
